doc/usage/restructuredtext/domains.rst

```diff
@@ -706,6 +706,72 @@ Inline Expressions and Types
    .. versionadded:: 3.0
 
 
+Namespacing
+~~~~~~~~~~~
+
+.. versionadded:: 3.1
+
+The C language it self does not support namespacing, but it can sometimes be
+useful to emulate it in documentation, e.g., to show alternate declarations.
+The feature may also be used to document members of structs/unions/enums
+separate from their parent declaration.
+
+The current scope can be changed using three namespace directives.  They manage
+a stack declarations where ``c:namespace`` resets the stack and changes a given
+scope.
+
+The ``c:namespace-push`` directive changes the scope to a given inner scope
+of the current one.
+
+The ``c:namespace-pop`` directive undoes the most recent
+``c:namespace-push`` directive.
```

Comment:
I think documentation is not a program. So I prefer not to add the push and pop mechanism into document. Is it natural to use the mechanism in the C world? If not, I think the declarative way; only using `c:namespace` directive is better.

Reply:
One reason is for consistency with the C++ domain, and while the languages are of course different, I see it more as a documentation tool. The situation I have used it in is to have a namespace declaration in the top of a document or in the beginning of an example. Then locally, e.g., for a struct, push to get into that scope and document members and pop afterwards. For example:
```rst
Example 1
---------

.. c:namespace:: @example_1

The datastructure can be document as

.. c:struct:: MyStruct

with the following members

.. c:namespace-push:: MyStruct

.. c:var:: int a

.. c:namespace-pop::

```

Basically, it offers convenience for the style of writing where nested declarations are written (semi-)independently of the parent declaration.

Reply:
IMO, using full namespace declaration is explicit than the push/pop approach.
```
Example 1
---------

.. c:namespace:: @example_1

...

.. c:namespace:: @example_1.MyStruct

...

.. c:namespace:: @example_1
```

Further speaking, it would be better if `c:struct` directive automatically modifies the namespace between processing.

Anyway, I did not know that the approach has been already added to the cpp domain. So either is okay to me.

Reply:
The ``c:struct`` directive is indeed tracking the scope, but that only applies for nested declarations. This is aimed at declarations inside the scope, but textually written somewhere else. While the full namespace declaration is indeed more explicit, it becomes very impractical when script-generated rst is involved or when the distance between the declarations is large.